pages/building-packages-for-runtimes.md

```diff
@@ -0,0 +1,239 @@
+} else {
+  module.exports = require("./cjs/react-router.js");
+}
+```
+
+This kinda of context and env aware of usage of built-ins. Will give better control for the bundlers to polyfill only the modules that are actually needed. Instead of polyfilling all the node built-ins every time a build is carried out for browsers.
+
+## JSPM
+
+We have seen multiple use cases for improving package compatibility within the ecosystem. Now, let's examine how `jspm` handles pre-existing packages in npm. It’s a well known issue when esm modules went into stable. The transition was not at all smooth in the eco-system. Because, for a project to work in js eco-system now.
+
+The build pipelines became so complex, it’s a proper combination of `bundlers`, `transpilers`, `node` and multiple module `formats`. Loading a `cjs` module from `npm` into project that is configuring to build as `esm` and vice-versa became so complex. At some point left most of the systems out of sync.
+
+`JSPM` builds all packages from `npm` ahead of time to spec complaint esm modules. And serves them using a distributed global CDN, regardless of the format in which the packages are published to npm. This makes it seamless for loading any module from `npm` and use it in any project at anytime.
+
+**Fields inside package.json**
+
+**browser**
+
+If a package is specifying [browser](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#browser) field in the package.json. And when these packages are used in generating `import-maps` for browser as target. Then jspm uses this field instead of `main`, as the package authors explicitly stating the module that needs to be used when loading the same build inside a browser environments.
+
+**module**
+
+Module build detection in jspm. Let’s say, we are exposing `cjs` build from `main` and `esm` version in `module` field. And we don’t have `type` as `module` set on the package. Nor, the entry point is ending with `.mjs` extension. Something like below
+
+```json
+"main": "dist/index.js",
+"module": "dist/index.es.js"
+```
+
+This is one of the most used pattern in publishing packages. The builder uses `main` entry and builds the module into `esm` . This is because the `module` field is not in official specification but mostly used by `bundlers`. JSPM always follows the node specification.
+
+**exports**
+
+The package-builder from jspm parses the packages and creates an `export-map` for those packages which don’t expose them by itself. In the process, the builder is even intelligent enough to detect [subpaths](https://nodejs.org/api/packages.html#subpath-patterns) depending on the usage of the module imports internally. Which in return makes the package exports more efficient.
+
+> If the package is exposing an `exports` map, JSPM uses it instead of re-generating the whole export map.
+
+```json
+"exports": {
+    ".": {
+      "import": "./dist/default/lib.mjs",
+      "require": "./dist/default/lib.js",
+      "umd": "./dist/default/lib.umd.js"
```

Comment:
I wouldn't generally encourage the `"umd"` key as it has no defined meaning or support anywhere I know of.